File: ErmisClient/Mobile/ermis_client/lib/client/common/chat_session.dart
/* Copyright (C) 2024 Ilias Koukovinis <ilias.koukovinis@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 * 
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

import 'dart:typed_data';
import 'message.dart';

class ChatSession {
  final int _chatSessionID;
  final int _chatSessionIndex;
  List<Member> _members;
  List<Message> _messages;
  bool _haveChatMessagesBeenCached;

  ChatSession(this._chatSessionID, this._chatSessionIndex)
      : _members = [],
        _messages = [],
        _haveChatMessagesBeenCached = false;

  ChatSession.withDetails(
      this._chatSessionID, this._chatSessionIndex, this._messages, this._members, this._haveChatMessagesBeenCached);

  void setMembers(List<Member> members) => _members = members;

  void setMessages(List<Message> messages) => _messages = messages;

  void setHaveChatMessagesBeenCached(bool haveChatMessagesBeenCached) =>
      _haveChatMessagesBeenCached = haveChatMessagesBeenCached;

  int get chatSessionID => _chatSessionID;

  int get chatSessionIndex => _chatSessionIndex;

  List<Member> get getMembers => _members;

  List<Message> get getMessages => _messages;

  bool get haveChatMessagesBeenCached => _haveChatMessagesBeenCached;

  @override
  int get hashCode => _chatSessionID.hashCode ^ _chatSessionIndex.hashCode;

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    if (other is! ChatSession) return false;
    return _chatSessionID == other._chatSessionID &&
        _chatSessionIndex == other._chatSessionIndex &&
        _haveChatMessagesBeenCached == other._haveChatMessagesBeenCached &&
        _members == other._members &&
        _messages == other._messages;
  }

  @override
  String toString() {
    return _members.map((member) => member.toString()).join(', ');
  }
}

class Member {
  String username;
  int clientID;
  Uint8List icon;
  bool isActive;

  Member(this.username, this.clientID, this.icon, this.isActive);

  void setUsername(String username) => this.username = username;
  void setClientID(int clientID) => this.clientID = clientID;
  void setIcon(Uint8List icon) => this.icon = icon;
  void setIsActive(bool isActive) => this.isActive = isActive;

  String get getUsername => username;
  int get getClientID => clientID;
  Uint8List get getIcon => icon;
  bool get getIsActive => isActive;

  @override
  int get hashCode => clientID.hashCode;

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    if (other is! Member) return false;
    return clientID == other.clientID &&
        icon == other.icon &&
        username == other.username &&
        isActive == other.isActive;
  }

  @override
  String toString() => '$username@$clientID';
}

File: ErmisClient/Desktop/src/github/koukobin/ermis/client/main/java/service/client/ChatSession.java
/* Copyright (C) 2023 Ilias Koukovinis <ilias.koukovinis@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 * 
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */
package github.koukobin.ermis.client.main.java.service.client;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.Objects;
import java.util.StringJoiner;

import github.koukobin.ermis.common.message_types.Message;

/**
 * @author Ilias Koukovinis
 *
 */
public class ChatSession {

	public static class Member {

		private String username;
		private int clientID;
		private byte[] icon;

		public Member() {}
		
		public Member(String username, int clientID, byte[] icon) {
			this.username = username;
			this.clientID = clientID;
			this.icon = icon.clone();
		}

		public void setUsername(String username) {
			this.username = username;
		}

		public void setClientID(int clientID) {
			this.clientID = clientID;
		}
		
		public void setIcon(byte[] icon) {
			this.icon = icon.clone();
		}
		
		public String getUsername() {
			return username;
		}

		public int getClientID() {
			return clientID;
		}
		
		public byte[] getIcon() {
			return icon.clone();
		}

		@Override
		public int hashCode() {
			return Objects.hash(clientID);
		}

		@Override
		public boolean equals(Object obj) {
			
			if (this == obj) {
				return true;
			}
			
			if (obj == null) {
				return false;
			}
			
			if (getClass() != obj.getClass()) {
				return false;
			}
			
			Member other = (Member) obj;
			return clientID == other.clientID 
					&& Arrays.equals(icon, other.icon)
					&& Objects.equals(username, other.username);
		}
		
		@Override
		public String toString() {
			return username + "@" + clientID;
		}
	}
	
	private final int chatSessionID;
	private final int chatSessionIndex; // Index of chat session that the server uses to access said chat session for the specific user
	
	private List<Member> members;
	private List<Message> messages;

	private boolean haveChatMessagesBeenCached;
	
	public ChatSession(int chatSessionID, int chatSessionIndex) {
		this.chatSessionID = chatSessionID;
		this.chatSessionIndex = chatSessionIndex;
		this.members = Collections.emptyList();
		this.messages = new ArrayList<>();
	}

	public ChatSession(int chatSessionID, int chatSessionIndex, List<Message> messages, List<Member> members, boolean haveChatMessagesBeenCached) {
		this.chatSessionID = chatSessionID;
		this.chatSessionIndex = chatSessionIndex;
		this.members = members;
		this.messages = messages;
		this.haveChatMessagesBeenCached = haveChatMessagesBeenCached;
	}

	public void setMembers(List<Member> members) {
		this.members = members;
	}

	public void setMessages(List<Message> messages) {
		this.messages = messages;
	}
	
	public void setHaveChatMessagesBeenCached(boolean haveChatMessagesBeenCached) {
		this.haveChatMessagesBeenCached = haveChatMessagesBeenCached;
	}
	
	public int getChatSessionID() {
		return chatSessionID;
	}

	public int getChatSessionIndex() {
		return chatSessionIndex;
	}

	public List<Member> getMembers() {
		return members;
	}

	public List<Message> getMessages() {
		return messages;
	}
	
	public boolean haveChatMessagesBeenCached()	{
		return haveChatMessagesBeenCached;
	}

	@Override
	public int hashCode() {
		return Objects.hash(chatSessionID, chatSessionIndex);
	}

	@Override
	public boolean equals(Object obj) {

		if (this == obj) {
			return true;
		}

		if (obj == null) {
			return false;
		}

		if (getClass() != obj.getClass()) {
			return false;
		}
		
		ChatSession other = (ChatSession) obj;
		return chatSessionID == other.chatSessionID 
				&& chatSessionIndex == other.chatSessionIndex
				&& haveChatMessagesBeenCached == other.haveChatMessagesBeenCached
				&& Objects.equals(members, other.members) 
				&& Objects.equals(messages, other.messages);
	}
	
	@Override
	public String toString() {
		StringJoiner joiner = new StringJoiner(", ");
		for (int i = 0; i < members.size(); i++) {
			joiner.add(members.get(i).toString());
		}
		return joiner.toString();
	}
}
